Added launchpad model in preparation of the next feature

kieranwoodrow opened the pull request.

kieranwoodrow committed:
Added launchpad model

smabhena reviewed (dismissed):
👍

bgouws commented on spaceR/spaceR/Models/Launchpads.swift:
```diff
@@ -0,0 +1,51 @@
+    let longitude: Double?
+    let launchAttempts: Int?
+    let launchSuccesses: Int?
+    let rocketsLaunched: [UUID?]
```
Hmm this seems a little bit strange, what type are you expecting here? 😄

kieranwoodrow replied:
This in an example return of the json

 "longitude": -120.5931438,
        "launch_attempts": 0,
        "launch_successes": 0,
        "rockets": [
            "5e9d0d95eda69955f709d1eb"
        ],

This is the schema declared in the backend
  "latitude": {
    "type": "Number",
    "default": null
  },
  "longitude": {
    "type": "Number",
    "default": null
  },
  "launch_attempts": {
    "type": "Number",
    "default": 0
  },
  "launch_successes": {
    "type": "Number",
    "default": 0
  },
  "rockets": [
    "UUID"
  ],

bgouws replied:
Hi @kieranwoodrow So that value is still just an array of `String` UUID just means that it will be a unique identifier, but you are not actually expecting a type of UUID.

bgouws replied:
Judging by this response you might also need to take another look at your model because  most of these need to decode into an object and not just an `Int` for example longitude will not decode into an `Double`

kieranwoodrow replied:
Okay, I changed it to String and not UUID, but i just tried to make an api call now and it returned and  mapped everything alright and i can access everything i need😅

bgouws replied:
Remember to push your code 😄

kieranwoodrow committed:
Added launchpad feature

You can now view all the launchpads and their respective title and picture. Next feature is to make the cards clickable to view the launchpad info such as the launch missions that took place there and all their respective info.

kieranwoodrow committed:
Shortened naming conventions

KirstyPallatte commented on spaceR/spaceR/Controllers/LaunchpadViewController.swift:
```diff
@@ -0,0 +1,68 @@
+        return cell
+    }
+    
+    func tableView(tableView: UITableView, didSelectRowAtIndexPath indexPath: IndexPath) {
```
Is this needed?

kieranwoodrow replied:
Removed this and implementing it in the next PR. (I was about to make a new feature and forgot to delete this before swicthing branches for the new feature. Good spot)

KirstyPallatte replied:
Perfect. 👍🏻

kieranwoodrow committed:
Removed tableview function. Will implement it in the next PR

KirstyPallatte reviewed (dismissed):
Looks good to me

bgouws commented on spaceR/spaceR/Controllers/LaunchpadViewController.swift:
```diff
@@ -0,0 +1,63 @@
+                self?.displayErrorAlert(title: .unsuccessfulLaunchpadApiCall,
+                                        errorMessage: .unsuccessfulLaunchpadApiCall,
+                                        buttonTitle: "Ok")
+                return
```
No need for this `return`

bgouws commented on spaceR/spaceR/CustomUI/UILaunchpadTableViewCell.swift:
```diff
@@ -0,0 +1,40 @@
+
+class UILaunchpadTableViewCell: UITableViewCell {
+    
+    @IBOutlet weak var launchpadTableViewTitle: UILabel!
```
These can be `private`

bgouws commented on spaceR/spaceR/CustomUI/UIRocketTableViewCell.swift:
```diff
@@ -29,7 +29,7 @@ class UIRocketTableViewCell: UITableViewCell {
     
     func rocketImageStyling() {
-        rocketTableViewImage.imageStyle()
+        rocketTableViewImage.imageStyle(type: "rocketTable")
```
Rather use an `enum` here 😄

kieranwoodrow replied:
cool, implemented it :)

bgouws commented on spaceR/spaceR/Models/Launchpads.swift:
```diff
@@ -0,0 +1,51 @@
+    let location: String?
+    let region: String?
+    let timezone: String?
+    let latitude: Double?
```
Just to be sure please make sure that there are not just decoding to nil 😃 also another note if you plan to use all this information great 😄 but if not then you don't need to decode the entire model.

kieranwoodrow replied:
Yeah i checked, they aren't. If the launchpad does return a nil because the property might be empty (like the launches at a launchpad if it a new launchpad), i handle them where i use them.
Also, i am planning on using all the info in the model :)

bgouws commented on spaceR/spaceR/ViewModels/LaunchpadViewModel.swift:
```diff
@@ -0,0 +1,69 @@
+        allLaunchpads = launchpads
+    }
+    
+    func getAllRockets() -> [Launchpads] {
```
This can be a computed property 😄 

`var rocketList: [Launchpads] { ...`

Please also consider the naming I used.

kieranwoodrow replied:
cool will do

bgouws commented on spaceR/spaceR/ViewModels/LaunchpadViewModel.swift:
```diff
@@ -0,0 +1,69 @@
+        return allLaunchpads
+    }
+    
+    func getLaunchpadCount() -> Int {
```
This can be a computed property 😄

kieranwoodrow replied:
cool will do

bgouws commented on spaceR/spaceR/ViewModels/LaunchpadViewModel.swift:
```diff
@@ -0,0 +1,69 @@
+        return launchpadImage
+    }
+    
+    func getLaunchpadTitle(index: Int) -> String {
```
Consider returning a full object for these functions below 😄

kieranwoodrow replied:
why? For these functions i pass in the index, then filter through to target the object in the array and then return that specific name/image etc for that object and then passing it back to the view controller. Hence no unpacking of the object in the controller) (it's working perfectly).
Why would it be better to return the object instead of the desired property?

bgouws replied:
Hi @kieranwoodrow so something to think about is how many times we are doing the same thing, so for example if you wanted to show this information on a cell for example you would have to do something like this: 


`viewModel.getLaunchpadTitle(index: indexPath.item)`
`viewModel.getLaunchpadLocation(index:  indexPath.item)`
`viewModel.getLaunchpadRegion(index: indexPath.item)`
`viewModel.getLaunchpadDetails(index: indexPath.item)`
and so on....

Whereas passing the entire object back keeps your code clean and readable and you would only need to unpack things in your cell.

If you are not using all this info you can remove these functions.

bgouws replied:
Let me know if that makes sense 😄

kieranwoodrow replied:
cool, removed some future functions

bgouws reviewed (changes requested):
Just some clean up

kieranwoodrow committed:
some cleanup

-Changed imageStyle function to accept enum type of table type rocket or launchpad (can extend it further when needed)
-Changed functions into computed variable where necessary
-Removed unnecessary return
-Made variables private

bgouws commented on spaceR/spaceR/Controllers/LaunchpadViewController.swift:
```diff
@@ -0,0 +1,62 @@
+            return UITableViewCell()
+        }
+        
+        cell.setLaunchpadCell(image: launchpadViewModel.getLaunchpadImage(index: indexPath.item),title: launchpadViewModel.getLaunchpadTitle(index: indexPath.item),  atIndex: indexPath.item)
```
Each parameter on its own line 😄 

```
cell.setLaunchpadCell(image: launchpadViewModel.getLaunchpadImage(index: indexPath.item),
                      title: launchpadViewModel.getLaunchpadTitle(index: indexPath.item),
                      atIndex: indexPath.item)
```

kieranwoodrow replied:
cool

bgouws reviewed (changes requested):
Couple more enhancements

kieranwoodrow committed:
some cleanup

bgouws reviewed (approved):
Looks good man 👍
